qubesadmin/base.py

```diff
@@ -176,72 +220,173 @@ def __getattr__(self, item):
+            return False
+
+        self.process_response(response.splitlines())
+        return True
+
+    # also called by get_all_data()
+    def process_response(self, lines):
+        '''Process data from the response to GetAll or GetAllData'''
+        self._clear_properties()
+        successful = True
+        for line in lines:
+            try:
+                (name, is_default, prop_type, value) = line.split(b'\t', 3)
+                name = name.decode('ascii')
+                prop_type = prop_type.decode('ascii')
+                is_default = is_default == b'D'
+
+                value = self._decode_value(value, prop_type)
+                if prop_type == 'str' and value is not None:
+                    value = unescape(value)
+
+                self._set_item(name, value, is_default)
+            except: # pylint: disable=bare-except
+                successful = False
+                traceback.print_exc(file=sys.stderr)
+
+        self._exhaustive = successful
+
+    @classmethod
+    def _local_properties(cls):
+        '''
+        Get set of property names that are properties on the Python object,
+        and must not be set on the remote object
+        '''
+        if "_local_properties_set" not in cls.__dict__:
+            props = set()
+            for class_ in cls.__mro__:
+                for key in class_.__dict__:
+                    props.add(key)
+            cls._local_properties_set = props
+
+        return cls._local_properties_set
 
     def __setattr__(self, key, value):
-        if key.startswith('_') or key in dir(self):
+        if key.startswith('_') or key in self._local_properties():
             return super(PropertyHolder, self).__setattr__(key, value)
         if value is qubesadmin.DEFAULT:
-            try:
-                self.qubesd_call(
-                    self._method_dest,
-                    self._method_prefix + 'Reset',
-                    key,
-                    None)
-            except qubesadmin.exc.QubesDaemonNoResponseError:
-                raise qubesadmin.exc.QubesPropertyAccessError(key)
+            self.__delattr__(key)
         else:
-            if isinstance(value, qubesadmin.vm.QubesVM):
-                value = value.name
-            if value is None:
-                value = ''
+            send_value = value
+            if isinstance(send_value, qubesadmin.vm.QubesVM):
+                send_value = value.name
+            if send_value is None:
+                send_value = ''
             try:
                 self.qubesd_call(
                     self._method_dest,
                     self._method_prefix + 'Set',
                     key,
-                    str(value).encode('utf-8'))
+                    str(send_value).encode('utf-8'))
             except qubesadmin.exc.QubesDaemonNoResponseError:
                 raise qubesadmin.exc.QubesPropertyAccessError(key)
 
+            self._set_item(key, value, False)
```

Comment:
This is fragile. Especially for long-living processes - anyone else can modify value in the meantime. But also any extension is free to modify value in event handler, including event caused by this very property set.
So, if do this at all, at least make sure the application also receive events. Or at least have it opt-in (and explicitly enable, only in short-lived processes, like *some* qvm-* tools).